### PR TITLE
Use torch.finfo settings for eps optionally

### PR DESCRIPTION
**Pull request by jstjohn**

### CODE REVIEW

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/train_geneformer.py [hunk 1]
```diff
@@ -147,7 +150,21 @@ def main(
...
+    precision_dtype = get_autocast_dtype(precision)
+    dtype_based_eps: float = torch.finfo(precision_dtype).eps
+    if layernorm_precision_based_eps:
+        layernorm_epsilon = dtype_based_eps
+    else:
+        layernorm_epsilon = 1e-12  # default from paper
+    if adam_precision_based_eps:
+        adam_epsilon = dtype_based_eps
+    else:
+        adam_epsilon = 1e-8  # default from adam settings
```
Comment (pstjohn) [inline sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/train_geneformer.py:166]: This is what the ESM-2 authors used

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/train_geneformer.py [hunk 2]
```diff
@@ -89,10 +89,19 @@ def main(
...
+    num_attention_heads: int = 4,
     log_every_n_steps: int = 50,
     gc_interval: int = 0,
     aligned_megatron_ddp: bool = False,
     recompilation_check: bool = False,
+    layernorm_eps: float = 1e-12,
+    adam_eps: float = 1e-8,
+    adam_w: float = 0.01,
+    adam_beta2: float = 0.999,
+    grad_reduce_in_fp32: bool = False,
```
Comment (sichu2023) [inline sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/train_geneformer.py:104]: We might want to switch to `adap_eps=1e-5` and `grad_reduce_in_fp32=True` in light of the wandb results.

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/train_geneformer.py [hunk 3]
```diff
@@ -621,7 +682,36 @@ def config_class_type(desc: str) -> Type[BioBertConfig]:
...
+    parser.add_argument(
+        "--layernorm-eps",
+        type=float,
+        default=1e-12,
+        help="Use data precision to define the epsilon for layer norm. Default is 1e-12 if not defined.",
+    )
+    parser.add_argument(
+        "--adam-eps",
+        type=float,
+        default=1e-8,
```
Comment (sichu2023) [inline sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/train_geneformer.py:694]: Similar to my comment above.

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/train_geneformer.py [hunk 4]
```diff
@@ -621,7 +682,36 @@ def config_class_type(desc: str) -> Type[BioBertConfig]:
...
+        type=float,
+        default=0.999,
+        help="Use data precision to define the epsilon for adam. Default is 0.999 if not defined.",
+    )
+    parser.add_argument(
+        "--grad-reduce-in-fp32",
+        action="store_true",
+        default=False,
+        help="Use higher precision fp32 grad reduction in ddp mode.",
+    )
```
Comment (sichu2023) [inline sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/train_geneformer.py:714]: Similar to above.

**Review by sichu2023 — APPROVED** [review]

**Pull request closed without merging**